airgen/src/lib.rs

```diff
@@ -56,15 +63,52 @@ pub fn compile(input: AnalysisASMFile) -> PILGraph {
         instances.push((location, ty));
     }
 
+    // count incoming permutations for each machine.
+    let mut incoming_permutations = instances
+        .iter()
+        .map(|(location, _)| (location.clone(), 0))
+        .collect();
+
     // visit the tree compiling the machines
-    let objects = instances
+    let mut objects: BTreeMap<_, _> = instances
         .into_iter()
         .map(|(location, ty)| {
-            let object = ASMPILConverter::convert_machine(&location, &ty, &input);
+            let object = ASMPILConverter::convert_machine(
+                &location,
+                &ty,
+                &input,
+                &mut incoming_permutations,
+            );
             (location, object)
         })
         .collect();
 
+    // add pil code for the selector array and related constraints
+    for (location, count) in incoming_permutations {
+        let obj = objects.get_mut(&location).unwrap();
+        if count > 0 {
+            // TODO: only block machines actually need a used flag to
```

Comment:
The `used` flag is only used to filter out dummy blocks in the caller, and those (AFAIU) are not an issue in VMs, no?

Reply:
For the issue of accessing the selectors, is accessing the used flag directly (i.e., sum of selectors) instead of the selector array enough?
We could do one of the following:

- a) try a builtin for that

- b) allow machines to have an optional declaration (`selector foo;`?), which would be the name for the used flag in place of the default "hidden" name, and thus would be referenceable by machine code.

What do you think? I kinda like the second option

Reply:
Ah, right, true!